core/arch/x86/emit_utils.c

```diff
@@ -1354,13 +1354,59 @@ append_restore_simd_reg(dcontext_t *dcontext, instrlist_t *ilist, bool absolute)
         int i;
         uint opcode = move_mm_reg_opcode(true /*align16*/, true /*align32*/);
         ASSERT(proc_has_feature(FEATURE_SSE));
-        for (i = 0; i < proc_num_simd_saved(); i++) {
+        instr_t *post_restore = NULL;
+        instr_t *pre_avx512_restore = NULL;
+        if (ZMM_ENABLED()) {
+            post_restore = INSTR_CREATE_label(dcontext);
+            pre_avx512_restore = INSTR_CREATE_label(dcontext);
+            APP(ilist,
+                INSTR_CREATE_mov_imm(
+                    dcontext, opnd_create_reg(SCRATCH_REG1),
+                    OPND_CREATE_INTPTR((ptr_int_t)&d_r_avx512_code_in_use)));
+            APP(ilist,
+                XINST_CREATE_load_1byte_zext4(dcontext, opnd_create_reg(SCRATCH_REG1),
+                                              OPND_CREATE_MEM8(SCRATCH_REG1, 0)));
+            APP(ilist,
+                INSTR_CREATE_test(dcontext, opnd_create_reg(SCRATCH_REG1),
+                                  opnd_create_reg(SCRATCH_REG1)));
+            APP(ilist,
+                INSTR_CREATE_jcc(dcontext, OP_jnz,
+                                 opnd_create_instr(pre_avx512_restore)));
+        }
+        for (i = 0; i < proc_num_simd_sse_avx_saved(); i++) {
             APP(ilist,
                 instr_create_1dst_1src(
                     dcontext, opcode, opnd_create_reg(REG_SAVED_XMM0 + (reg_id_t)i),
                     OPND_DC_FIELD(absolute, dcontext, OPSZ_SAVED_XMM,
                                   SIMD_OFFSET + i * MCXT_SIMD_SLOT_SIZE)));
         }
+        if (ZMM_ENABLED()) {
+            APP(ilist, INSTR_CREATE_jmp(dcontext, opnd_create_instr(post_restore)));
+            APP(ilist, pre_avx512_restore /*label*/);
+            uint opcode_avx512 = move_mm_avx512_reg_opcode(true /*align64*/);
+            for (i = 0; i < proc_num_simd_registers(); i++) {
+                APP(ilist,
+                    instr_create_1dst_2src(
+                        dcontext, opcode_avx512,
+                        opnd_create_reg(DR_REG_START_ZMM + (reg_id_t)i),
+                        opnd_create_reg(DR_REG_K0),
+                        OPND_DC_FIELD(absolute, dcontext, ZMM_REG_SIZE,
+                                      SIMD_OFFSET + i * MCXT_SIMD_SLOT_SIZE)));
+            }
+            /* Moving 64-bit masks is not supported in 32-bit mode. */
+            bool has_avx512bw = IF_X64_ELSE(proc_has_feature(FEATURE_AVX512BW), false);
```

Comment:
For 32-bit will the feature still show up?

Reply:
Yes. That said, I think this can be considered a bug. I'm not sure how you're supposed to save the 64-bit mask register value in 32-bit. The k* AVX-512 VEX instructions can't handle it. Ok with you to add a comment and describe, along with a SYSLOG_WARNING pointing out that we are saving only a WORD?

Reply:
So xsave is the only way to save the full mask registers in 32-bit mode?  Hmm.  If we want to punt on this for now maybe file a separate issue with a FIXME here for it.  We may want to explore/measure xsave in any case.

Reply:
SYSLOG_WARNING? Would be hard to debug if we forget about it.

Reply:
For some reason I thought the k* instructions do not support memory. There actually is

```
VEX.L0.0F.W1 90 /r
KMOVQ k1, k2/m64
```

It is implemented correctly in the decode tables. Fixed. The sequence is using this instruction now.